File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    answer = n
    tree = [[] for _ in range(n+1)]
    visited = []
    for w in wires:
        tree[w[0]].append(w[1])
        tree[w[1]].append(w[0])
        
    def getNumOfNode(x, v1, v2):
        q = deque([x])
        visited[x] = True
        cnt = 0
        while q:
            curr = q.popleft()
            cnt+=1
            for nxt in tree[curr]:
                if not visited[nxt]:
                    if nxt not in [v1, v2]:
                        q.append(nxt)
                        visited[nxt] = True
        return cnt
    
    for v1 in range(1, n+1):
        for v2 in range(v1+1, n+1):
            if v2 in tree[v1]: #v1과 v2가 연결되어 있다면, 연결을 끊고 각 영역의 송전탑 개수의 차이를 구함
                visited = [False for _ in range(n+1)]
                cnt1 = getNumOfNode(v1, v1, v2)
                visited = [False for _ in range(n+1)]
                cnt2 = getNumOfNode(v2, v1, v2)
                answer = min(answer, abs(cnt1-cnt2))
    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        HashMap<Integer, ArrayList<Integer>> tree = new HashMap<>();
        int answer = n+1;
        
        for(int i=1; i<=n; i++) {
            ArrayList<Integer> list = new ArrayList<>();
            tree.put(i, list);
        }
        
        for(int[] w: wires){
            ArrayList<Integer> l1 = new ArrayList<>();
            ArrayList<Integer> l2 = new ArrayList<>();
            
            l1 = tree.get(w[0]);
            l1.add(w[1]);
            l2 = tree.get(w[1]);
            l2.add(w[0]);
        }
        
        //연결을 하나씩 끊어보고 각각 차이를 비교
        for(int[] w: wires){
            
            //1. w[0], w[1]의 연결 끊음
            ArrayList<Integer> l1 = new ArrayList<>();
            ArrayList<Integer> l2 = new ArrayList<>();
            
            l1 = tree.get(w[0]);
            l2 = tree.get(w[1]);
            if(l1.contains(w[1])) l1.remove(Integer.valueOf(w[1]));
            if(l2.contains(w[0])) l2.remove(Integer.valueOf(w[0]));
            
            //2. w[0]과 w[1]을 시작점으로 하는 트리 크기 구함
            int t1_size = treeSize(tree, w[0]);
            int t2_size = treeSize(tree, w[1]);
            answer = Math.min(answer, Math.abs(t1_size- t2_size));
            

            //3. 원상 복구하기
            l1.add(w[1]);
            l2.add(w[0]);           
            
        }
        
        
        return answer;
    }
    
    //주어진 트리의 크기 구하는 함수
    private int treeSize(HashMap<Integer, ArrayList<Integer>> tr, int start){
        Queue<Integer> q = new LinkedList<>();
        List<Integer> visited = new ArrayList<>();
        q.offer(start);
        visited.add(start);
        
        while(!q.isEmpty()){
            int curr = q.poll();
            for(int nxt: tr.get(curr)){
                if(!visited.contains(nxt)){
                    q.offer(nxt);
                    visited.add(nxt);
                }
            }
        }
        return visited.size();
        
    }
}